tests/test_olio_grid_functions.py

```diff
@@ -0,0 +1,236 @@
+        result_exit_error = left_right_foursome(full_pillar_list = full_p_list_exit_error, p_index = p_index)
+        assert "code failure whilst taking exit sides from dubious full pillar list" in str(excinfo.value)
+
+
+def test_resequence_nexus_corp():
+
+    # TODO: confirm you used the correct cornerpoint ordering from Nexus (pg 1424 Nexus Keyword file)
+    # --------- Arrange ----------
+    corner_points = np.array([[[[[[[0., 0., 0.], [10., 0., 0.]], [[10., -10., 0.], [0., -10., 0.]]],
+                                 [[[0., 0., 10.], [10., 0., 10.]], [[10., -10., 10.], [0., -10., 10.]]]]]]])
+
+    expected_result = np.array([[[[[[[0., 0., 0.], [10., 0., 0.]], [[0., -10., 0.], [10., -10., 0.]]],
+                                   [[[0., 0., 10.], [10., 0., 10.]], [[0., -10., 10.], [10., -10., 10.]]]]]]])
+
+    # --------- Act ----------
+    resequence_nexus_corp(corner_points = corner_points)
+
+    # --------- Assert ----------
+    np.testing.assert_almost_equal(corner_points, expected_result)
+
+
+# TODO: Find out if grid is too simple
+def test_random_cell(tmp_path):
+
+    # --------- Arrange----------
+    epc = os.path.join(tmp_path, 'grid.epc')
+    model = rq.new_model(epc)
+
+    # create a basic block grid
+    dxyz = (55.0, 65.0, 27.0)
+    grid = grr.RegularGrid(model, extent_kji = (4, 3, 2), title = 'concrete', origin = (0.0, 0.0, 1000.0), dxyz = dxyz)
+    grid.create_xml(add_cell_length_properties = True)
+    corner_points = grid.corner_points()
+
+    # --------- Act----------
+    (k, j, i) = random_cell(corner_points = corner_points)
+
+    # --------- Assert----------
+    assert k <= 4
+    assert j <= 3
+    assert i <= 2
+
+
+def test_triangles_for_cell_faces():
+    # --------- Arrange----------
+    corner_points = np.array([[[[0., 0., 0.], [10., 0., 0.]], [[0., -10., 0.], [10., -10., 0.]]],
+                              [[[0., 0., 10.], [10., 0., 10.]], [[0., -10., 10.], [10., -10., 10.]]]])
+    # --------- Act----------
+    tri = triangles_for_cell_faces(cp = corner_points)
+    # face centre points
+    k_face_centre_points = [np.array([5., -5., 0.]), np.array([5., -5., 10])]
+    j_face_centre_points = [np.array([5., 0., 5.]), np.array([5., -10., 5.])]
+    i_face_centre_points = [np.array([0., -5., 5.]), np.array([10., -5., 5.])]
+
+    # --------- Assert----------
+    np.testing.assert_almost_equal(tri[0, :, :, 0][0][0], k_face_centre_points[0])
+    np.testing.assert_almost_equal(tri[0, :, :, 0][1][0], k_face_centre_points[1])
+    np.testing.assert_almost_equal(tri[1, :, :, 0][0][0], j_face_centre_points[0])
+    np.testing.assert_almost_equal(tri[1, :, :, 0][1][0], j_face_centre_points[1])
+    np.testing.assert_almost_equal(tri[2, :, :, 0][0][0], i_face_centre_points[0])
+    np.testing.assert_almost_equal(tri[2, :, :, 0][1][0], i_face_centre_points[1])
+
+
+def test_actual_pillar_shape(tmp_path):  # TODO: test with more complicated grid geometries
```

Comment:
Yes, please add tests to cover the two other possible outcomes: 'straight' (but not vertical) and 'curved'.

I can give some suggestions about setting up other geometries – let's discuss.